src/parser.rs

```diff
@@ -1020,4 +1058,55 @@ CC,featureC
             ]
         );
     }
+
+    #[test]
+    fn test_cache() {
+        // Test for perfect vowpal-compatible hashing
+        let vw_map_string = r#"
+AA,featureA
+BB,featureB
+CC,featureC
+"#;
+        let vw = vwmap::VwNamespaceMap::new(vw_map_string).unwrap();
+
+        fn str_to_cursor(s: &str) -> Cursor<Vec<u8>> {
+            Cursor::new(s.as_bytes().to_vec())
+        }
+
+        let mut rr = VowpalParser::new(&vw);
+
+        let mut buf = str_to_cursor("|BB b |AA:3 a:2.0 \n");
+        let buf_result =
+            [8, 255, 1065353216, 2147876872, 1123906636, 2147483648, 292540976, 1086324736];
+
+        assert_eq!(
+            rr.next_vowpal(&mut buf).unwrap(),
+            buf_result
+        );
+
+        let cache_input_str = "|BB b \n";
+        let mut cache_buf = str_to_cursor(cache_input_str);
+        let (cache_result, cache_result_size) = rr.next_vowpal_with_size(&mut cache_buf).unwrap();
+
+        assert_eq!(
+            cache_result,
+            [
```

Comment:
Consider adding a test that is actually representative of how instances look like, e.g., a label + a handful of namespaces, including multivalues

Reply:
I added it, but decided to remove it in the end as it is very hard to maintain. Would prefer that we add an integration test that checks the full flow. We can discuss & add it as part of a separate PR as this one is quite large.

Reply:
Agreed, an integration test/CI part for serving is certainly on the roadmap then

Reply:
completely agree with @SkBlaz , this is important